### PR TITLE
bpo-39244: multiprocessing get all start methods

### PR DESCRIPTION
**Pull request by idomic**
https://bugs.python.org/issue39244

### CODE REVIEW

**Comment by idomic** [comment]
@taleinat clean PR

**Comment by idomic** [comment]
Adding @taleinat , do I have to add tests for this?

**Comment by idomic** [comment]
ping @taleinat

File: Lib/multiprocessing/context.py [hunk 1]
```diff
@@ -256,6 +256,8 @@ def get_start_method(self, allow_none=False):
     def get_all_start_methods(self):
         if sys.platform == 'win32':
             return ['spawn']
+        elif sys.platform == 'darwin':
+            return ['spawn', 'fork', 'forkserver']
```
Comment (taleinat) [inline Lib/multiprocessing/context.py]: Does this not require the `if reduction.HAVE_SEND_HANDLE:` check as below?

**Comment by taleinat** [comment]
> Adding @taleinat , do I have to add tests for this?

The test already included seems sufficient.

**Comment by taleinat** [comment]
> ping @taleinat

Sorry, I'm extremely busy these days...

File: Lib/multiprocessing/context.py (hunk 1 above)
Comment (idomic) [reply Lib/multiprocessing/context.py]: Moved the platform check to be included in the HAVE_SEND_HANDLE if block

**Comment by idomic** [comment]
> > ping @taleinat
> 
> Sorry, I'm extremely busy these days...

That's alright, don't worry about it and stay safe!

File: Lib/multiprocessing/context.py [hunk 2]
```diff
@@ -258,7 +258,10 @@ def get_all_start_methods(self):
             return ['spawn']
         else:
             if reduction.HAVE_SEND_HANDLE:
-                return ['fork', 'spawn', 'forkserver']
+                if sys.platform == 'darwin':
```
Comment (taleinat) [inline Lib/multiprocessing/context.py]: Are you certain `reduction.HAVE_SEND_HANDLE` will always be true on macOS? If not 100% this is true on older and future versions, then a similar change should be done in the `else` clause below.

File: Lib/test/_test_multiprocessing.py [hunk 3]
```diff
@@ -5022,6 +5022,8 @@ def test_get_all(self):
         methods = multiprocessing.get_all_start_methods()
         if sys.platform == 'win32':
             self.assertEqual(methods, ['spawn'])
+        elif sys.platform == 'darwin':
+            self.assertEqual(methods, ['spawn', 'fork', 'forkserver'])
```
Comment (taleinat) [inline Lib/test/_test_multiprocessing.py]: ISTM this should allow 'forkserver' to be missing, as done by the assertion two lines below.

**Review by taleinat — CHANGES_REQUESTED** [review]

File: Lib/test/_test_multiprocessing.py (hunk 3 above)
Comment (idomic) [reply Lib/test/_test_multiprocessing.py]: Dropped this test, I agree it's redundant + fixed the code to have the return value in the else clause

**Comment by idomic** [comment]
I have made the requested changes; please review again

File: Lib/multiprocessing/context.py [hunk 4]
```diff
@@ -257,10 +257,11 @@ def get_all_start_methods(self):
         if sys.platform == 'win32':
             return ['spawn']
         else:
+            if sys.platform == 'darwin':
+                return ['spawn', 'fork', 'forkserver']
```
Comment (taleinat) [inline Lib/multiprocessing/context.py]: This breaks the intention of the existing code: The purpose of the `HAVE_SEND_HANDLE` check is to decide whether to include 'forkserver' as an option.  It should be done regardless of platform.

Perhaps just add a conditional after this code the brings 'spawn' to the beginning of the list on macOS.

File: Misc/NEWS.d/next/Library/2020-02-23-15-09-47.bpo-39244.aBK5IM.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Fixed :class:`multiprocessing.context.get_all_start_methods`
+If the context is macOS, return the default methods by order.
```
Comment (taleinat) [inline Misc/NEWS.d/next/Library/2020-02-23-15-09-47.bpo-39244.aBK5IM.rst]: This wording is imprecise; this should always return the default method first, which was not done on the macOS platform.

**Review by taleinat — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2020-02-23-15-09-47.bpo-39244.aBK5IM.rst (hunk 5 above)
Comment (idomic) [reply Misc/NEWS.d/next/Library/2020-02-23-15-09-47.bpo-39244.aBK5IM.rst]: Done.

File: Lib/multiprocessing/context.py [hunk 6]
```diff
@@ -259,8 +259,9 @@ def get_all_start_methods(self):
         else:
             if reduction.HAVE_SEND_HANDLE:
```
Comment (taleinat) [inline Lib/multiprocessing/context.py:260]: This still needs to return a list without 'forkserver' when `HAVE_SEND_HANDLE` is false.

```suggestion
            methods = ['spawn', 'fork'] if sys.platform == 'darwin' else ['fork', 'spawn']
            if reduction.HAVE_SEND_HANDLE:
                methods.append('forkserver')
            return methods
                
            if sys.platform == 'darwin':
```

File: Lib/multiprocessing/context.py [hunk 7]
```diff
@@ -259,8 +259,9 @@ def get_all_start_methods(self):
...
                 return ['fork', 'spawn', 'forkserver']
-            else:
-                return ['fork', 'spawn']
+            if sys.platform == 'darwin':
+                return ['spawn', 'fork', 'forkserver']
```
Comment (taleinat) [inline Lib/multiprocessing/context.py]: This still needs to return a list without 'forkserver' when `HAVE_SEND_HANDLE` is false.

```suggestion
            methods = ['spawn', 'fork'] if sys.platform == 'darwin' else ['fork', 'spawn']
            if reduction.HAVE_SEND_HANDLE:
                methods.append('forkserver')
            return methods
```

**Review by taleinat — APPROVED** [review]
LGTM

**Pull request merged**